pkg/notify/notifier/wechat/msg.go

```diff
@@ -0,0 +1,296 @@
+
+func init() {
+	notify.Register("Wechat", NewWechatNotifier)
+}
+
+func NewWechatNotifier(logger log.Logger, val interface{}, opts *nmv1alpha1.Options) notify.Notifier {
+
+	receivers, ok := val.([]*notifyconfig.Wechat)
+	if !ok {
+		_ = level.Error(logger).Log("msg", "Notifier: value type error")
+		return nil
+	}
+
+	notifier := &Notifier{
+		wechat:  make(map[string]*config.WechatConfig),
+		logger:  logger,
+		timeout: DefaultSendTimeout,
+		client:  ats.client,
+	}
+
+	if opts != nil && opts.NotificationTimeout != nil && opts.NotificationTimeout.Wechat != nil {
+		notifier.timeout = time.Second * time.Duration(*opts.NotificationTimeout.Wechat)
+	}
+
+	for _, receiver := range receivers {
+		c := notifier.clone(receiver.WechatConfig)
+		key, err := notifier2.Md5key(c)
+		if err != nil {
+			_ = level.Error(logger).Log("msg", "WechatNotifier: get notifier error", "error", err.Error())
+			continue
+		}
+
+		w, ok := notifier.wechat[key]
+		if !ok {
+			w = c
+		}
+
+		if len(receiver.ToUser) > 0 {
+			w.ToUser += "|" + receiver.ToUser
+		}
+		w.ToUser = strings.TrimPrefix(w.ToUser, "|")
+
+		if len(receiver.ToTag) > 0 {
+			w.ToTag += "|" + receiver.ToTag
+		}
+		w.ToTag = strings.TrimPrefix(w.ToTag, "|")
+
+		if len(receiver.ToParty) > 0 {
+			w.ToParty += "|" + receiver.ToParty
+		}
+		w.ToParty = strings.TrimPrefix(w.ToParty, "|")
+
+		notifier.wechat[key] = w
+	}
+
+	return notifier
+}
+
+func (n *Notifier) Notify(data template.Data) []error {
+
+	var errs []error
+	send := func(c *config.WechatConfig) (bool, error) {
+		ctx, cancel := context.WithTimeout(context.Background(), n.timeout)
+		defer cancel()
+
+		msg := &weChatMessage{
+			Text: weChatMessageContent{
+				Content: n.getMsg(data),
+			},
+			ToUser:  c.ToUser,
+			ToParty: c.ToParty,
+			Totag:   c.ToTag,
+			AgentID: c.AgentID,
+			Type:    "text",
+			Safe:    "0",
+		}
+
+		var buf bytes.Buffer
+		if err := jsoniter.NewEncoder(&buf).Encode(msg); err != nil {
+			_ = level.Error(n.logger).Log("msg", "WechatNotifier: encode error", "error", err.Error())
+			return false, err
+		}
+
+		postMessageURL := c.APIURL.Copy()
+		postMessageURL.Path += "message/send"
+		q := postMessageURL.Query()
+		res := make(chan interface{})
+		ats.get(c, ctx, res)
+		accessToken := ""
+		select {
+		case <-ctx.Done():
+			_ = level.Error(n.logger).Log("msg", "WechatNotifier: get accesstoken timeout")
+			return true, fmt.Errorf("get accesstoken timeout")
+		case val := <-res:
+			switch val.(type) {
+			case error:
+				_ = level.Error(n.logger).Log("msg", "WechatNotifier: get accesstoken error", "error", val.(error).Error())
+				return true, val.(error)
+			case string:
+				accessToken = val.(string)
+			}
+		}
+
+		q.Set("access_token", accessToken)
+		postMessageURL.RawQuery = q.Encode()
+
+		req, err := http.NewRequest(http.MethodPost, postMessageURL.String(), &buf)
+		if err != nil {
+			_ = level.Error(n.logger).Log("msg", "WechatNotifier: wechat notify error", "error", err.Error())
```

Comment:
many logs in this function are the same "WechatNotifier: wechat notify error", maybe we can make these logs more specific to reflect each error case